stan/math/rev/functor/reduce_sum.hpp

```diff
@@ -38,24 +38,36 @@ struct reduce_sum_impl<ReduceFunction, require_var_t<ReturnType>, ReturnType,
    */
   struct recursive_reducer {
     const size_t num_vars_per_term_;
-    const size_t num_vars_shared_terms_;  // Number of vars in shared arguments
-    double* sliced_partials_;  // Points to adjoints of the partial calculations
+    const size_t num_vars_shared_terms_;  // Number of vars in shared
+                                          // arguments
+    vari** varis_;      // Points to array of the input operands
+    double* partials_;  // Points to adjoints of the partial
+                        // calculations
     Vec vmapped_;
     std::ostream* msgs_;
     std::tuple<Args...> args_tuple_;
+    const std::thread::id main_thread_;
+    std::unique_ptr<double[]> args_adjoints_holder_;
```

Comment:
What is args_adjoints_holder_ used for?

Reply:
To store a pointer to the double array for the shared adjoints. The root reducer writes directly to the partials array and the child reducers allocate their own space for that. Hence we need eigen maps everywhere is what I figured.

Reply:
But can the `Eigen::Map` just store that pointer directly?

Reply:
Don't think so. Eigen Map is there to deal with raw memory buffers... but the unique_ptr is there to cleanup things. The Map alone would not de-allocate the double array.